src/ejabberd_receiver.erl

```diff
@@ -136,15 +140,24 @@ init([Socket, SockMod, Shaper, MaxStanzaSize]) ->
 %%                                      {stop, Reason, State}
 %% Description: Handling call messages
 %%--------------------------------------------------------------------
-handle_call({starttls, TLSSocket}, _From, State) ->
-    StateAfterReset = reset_parser(State),
-    NewState = StateAfterReset#state{socket = TLSSocket,
-                                 sock_mod = fast_tls},
-    case fast_tls:recv_data(TLSSocket, <<"">>) of
-        {ok, TLSData} ->
-            {reply, ok, process_data(TLSData, NewState), ?HIBERNATE_TIMEOUT};
-        {error, _Reason} ->
-            {stop, normal, ok, NewState}
+handle_call(get_socket, _From, #state{socket = Socket} = State) ->
+    {reply, {ok, Socket}, State, ?HIBERNATE_TIMEOUT};
+handle_call({starttls, TLSOpts}, From, #state{parser = Parser,
+                                              socket = TCPSocket} = State) ->
+    deactivate_socket(State),
+    gen_server:reply(From,ok), %% let ejabberd_socket finish tcp negotiations
```

Comment:
`, `.
And I still don't know why we need to reply.
Actually, If c2s would want to write something funny, it can interfere with work of ssl.

Reply:
we need to reply to let ejabberd_c2s process (ejabberd_socket module), send the last chunk to client before the ssl handshake. ssl handshake always starts from the client's request, so we need to tell client that we are ready and waiting for his request.

Reply:
How do we know, that reply was received by ejabberd_c2s and *processed*?

Reply:
Actually is there a reason why `c2s` can't send `proceed` element before initiating StartTLS handshake? Does it have to call `:starttls` function first?

Reply:
@fenek  yes, there is the reason. socket must be deactivated first! other-way first messages of the handshake can go to the wrong process (in case of standard ssl backend). 
@arcusfelis sorry, don't get your question.

Reply:
I think that @arcusfelis' concern is valid. Can you guarantee that caller process (c2s) will manage to send `<proceed/>` before `ejabberd_tls:tcp_to_tls` call?

Reply:
it doesn't matter, tls handshake always starts from clients request. 

so the first action of ejabberd_tls:tcp_to_tls (for just_tls backend) is to wait for client's request.

for fast_tls it's not a problem as well. but usually the real handshake occurs on ```ejabberd_tls:recv_data(TLSSocket, <<"">>)``` it is a bit tricky but it's how fast_tls works.

so generally this approach is 100% correct

Reply:
OK, I understand. Thanks. :)